### PR TITLE
Auth: Add storage volume and bucket location to URL in access check

### LINKED ISSUE #13365: Auth: Use correct location of storage volume/bucket when performing access check
See https://github.com/canonical/lxd/pull/13364#discussion_r1574398949_

When performing an access check on a storage volume, we need to include it's location in the URL, otherwise the permission checker will not recognise it.

### PR DESCRIPTION
**Pull request by markylaing**
This PR fixes fine grained authorization for storage volumes and storage buckets in clustered LXD. 

When listing volumes, we can use the `Location` field of the storage volume or bucket to perform an access check.

When performing an action on a single storage volume, we need to perform the following checks:
1. Is the pool containing the volume remote? If so, the `target` query parameter may have been used to target a particular cluster member, but should not be used in the URL for the volume (volumes in remote pools do not have a location).
2. Is the volume located on another cluster member? If so, and the target parameter is set, use the target parameter as the location. If so, and the target parameter is unset, use the cluster member name as the location. If not, use the server name as the location.

To prevent extra queries from being performed when a request is forwarded to another node. Each node will assume that forwarded requests have already reached their intended destination. This is done by checking for `request.CtxForwardedProtocol` in the request context (I tried this originally by setting the `target` parameter but this broke storage volume migrations).

Closes #13365

### CODE REVIEW

File: lxd/storage_volumes.go [hunk 1]
```diff
@@ -1492,24 +1492,14 @@ func storagePoolVolumePost(d *Daemon, r *http.Request) response.Response {
...
 			}
 
-			srcPool, err := storagePools.LoadByName(s, srcPoolName)
-			if err != nil {
-				return response.SmartError(err)
-			}
-
-			if srcPool.Driver().Info().Name == "ceph" {
+			// FIXME: Should this be `if pool.Driver().Info().Remote {...}`? E.g. what about powerflex?
+			if pool.Driver().Info().Name == "ceph" {
```
Comment (markylaing) [inline lxd/storage_volumes.go:1495]: @tomponline @roosterfish I thought this line was quite suspicious. Do you know if this block is specific to the Ceph driver? Or should it be for remote drivers?

File: lxd/storage_volumes_backup.go [hunk 2]
```diff
@@ -191,31 +190,29 @@ func storagePoolVolumeTypeCustomBackupsGet(d *Daemon, r *http.Request) response.
...
-		poolID, _, _, err = tx.GetStoragePool(ctx, poolName)
-
-		return err
-	})
+	pool, err := request.GetCtxValue[storagePools.Pool](r.Context(), ctxStoragePool)
 	if err != nil {
 		return response.SmartError(err)
 	}
 
+	// FIXME: Is handler supposed to forward requests if the target is remote?
```
Comment (markylaing) [inline lxd/storage_volumes_backup.go:198]: I was confused here as to why this handler forwards requests to other cluster members if the volume is remote but not when the target parameter is remote. Added the FIXME here so we can discuss before merging.

File: lxd/storage_volumes_backup.go [hunk 3]
```diff
@@ -336,27 +333,25 @@ func storagePoolVolumeTypeCustomBackupsPost(d *Daemon, r *http.Request) response
...
-	})
+	pool, err := request.GetCtxValue[storagePools.Pool](r.Context(), ctxStoragePool)
 	if err != nil {
 		return response.SmartError(err)
 	}
 
-	resp = forwardedResponseIfVolumeIsRemote(s, r, poolName, projectName, volumeName, cluster.StoragePoolVolumeTypeCustom)
-	if resp != nil {
-		return resp
+	// FIXME: Forward if target parameter is remote?
```
Comment (markylaing) [inline lxd/storage_volumes_backup.go:341]: Again here there doesn't seem to be any forwarding if the target is not this cluster member, but there is forwarding if the volume is located on another member.

File: lxd/storage_buckets.go [hunk 4]
```diff
@@ -602,6 +592,7 @@ func storagePoolBucketPut(d *Daemon, r *http.Request) response.Response {
 			return response.SmartError(err)
 		}
 
+		// FIXME: ETag check?
```
Comment (markylaing) [inline lxd/storage_buckets.go:595]: There doesn't appear to be an eTag check for updating a storage bucket.

**Review by markylaing — COMMENTED** [review]
I've added a few comments on things I noticed while working on this that might be worth revisiting.

File: lxd/storage_volumes.go (hunk 1 above)
Comment (roosterfish) [reply lxd/storage_volumes.go:1495]: Interesting. It doesn't look Ceph specific to me.
Did a quick search on the code base and also found https://github.com/canonical/lxd/blob/main/lxd/instance_post.go#L987. Looks all of those help moving Ceph based volumes from an offline cluster member. I agree this should be generic for remote storage drivers. 

Should we have a separate issue so we can follow up on this?

Comment (roosterfish) [reply lxd/storage_volumes.go:1495]: At least I cannot see why this logic should only apply for Ceph based storage vols.

Comment (tomponline) [reply lxd/storage_volumes.go:1495]: > Interesting. It doesn't look Ceph specific to me.

Agreed, it should be changed to cover any remote pool - although it will need to be tested with powerflex.

Comment (tomponline) [reply lxd/storage_volumes.go:1495]: > Did a quick search on the code base and also found https://github.com/canonical/lxd/blob/main/lxd/instance_post.go#L987. Looks all of those help moving Ceph based volumes from an offline cluster member. I agree this should be generic for remote storage drivers.
> 
> Should we have a separate issue so we can follow up on this?

Yes that looks like it shouldnt be ceph specific either.

I think a separate issue and a search for `== "ceph"` to check for the missed items from when the powerflex driver was added will be needed here.

@roosterfish presumably that means that some functionality in the powerflex driver isn't working correctly at this time?

Comment (roosterfish) [reply lxd/storage_volumes.go:1495]: @tomponline for this specific scenario that is possible yes. I'll have a look into this.

File: lxd/storage_volumes_backup.go (hunk 2 above)
Comment (markylaing) [reply lxd/storage_volumes_backup.go:198]: I've added an issue to track this here: https://github.com/canonical/lxd/issues/13548

File: lxd/storage_volumes_backup.go (hunk 3 above)
Comment (markylaing) [reply lxd/storage_volumes_backup.go:341]: I've added an issue to track this here: https://github.com/canonical/lxd/issues/13548

File: lxd/storage_buckets.go (hunk 4 above)
Comment (markylaing) [reply lxd/storage_buckets.go:595]: I've added an issue to track this here: https://github.com/canonical/lxd/issues/13549

**Comment by tomponline** [comment]
Please can you rebase

**Comment by markylaing** [comment]
@tomponline rebased and ready for review

File: shared/entity/type.go [hunk 5]
```diff
@@ -221,11 +221,8 @@ func (t Type) URL(projectName string, location string, pathArguments ...string)
 		u = u.WithQuery("project", projectName)
 	}
 
-	// Always set location if provided.
-	if location != "" {
-		u = u.WithQuery("target", location)
-	}
-
+	// Always set location if provided (empty or "none" locations are ignored).
+	u = u.Target(location)
```
Comment (tomponline) [inline shared/entity/type.go:225]: Does this affect any end-user API output or is this just internal?
Also if "(empty or "none" locations are ignored)" why is this needed?

File: lxd/storage_volumes.go [hunk 6]
```diff
@@ -2727,3 +2727,84 @@ func createStoragePoolVolumeFromBackup(s *state.State, r *http.Request, requestP
 	revert.Success()
 	return operations.OperationResponse(op)
 }
+
+// getRemoteVolumeNodeInfo figures out the cluster member on which the volume with the given name is defined. If it is
+// the local cluster member it returns nil and no error. If it is another cluster member it returns a db.NodeInfo containing
+// the name and address of the remote member. If there is more than one cluster member with a matching volume name, an
+// error is returned.
+func getRemoteVolumeNodeInfo(s *state.State, poolName string, projectName string, volumeName string, volumeType int) (*db.NodeInfo, error) {
```
Comment (tomponline) [inline lxd/storage_volumes.go:2852]: New DB functions should be taking a context arg so we can pass it to queries to allow cancellation in my view.

File: lxd/storage_buckets.go [hunk 7]
```diff
@@ -55,6 +55,54 @@ var storagePoolBucketKeyCmd = APIEndpoint{
...
+		}
+
+		request.SetCtxValue(r, ctxStoragePool, pool)
+
+		bucketName, err := url.PathUnescape(mux.Vars(r)["bucketName"])
+		if err != nil {
+			return response.SmartError(err)
+		}
+
+		// If the storage pool is a remote driver, the URL of the storage bucket will not contain a target parameter
```
Comment (tomponline) [inline lxd/storage_buckets.go:90]: This comment seems misleading, a client can add a target parameter to any request it pleases, so to say a URL "will not" contain a target parameter in certain scenarios doesn't seem accurate.

Perhaps "should not" would be clearer?

**Comment by tomponline** [comment]
> When performing an action on a single storage volume, we need to perform the following checks:
> 
>     1. Is the pool containing the volume remote? If so, the `target` query parameter may have been used to target a particular cluster member, but should not be used in the URL for the volume (volumes in remote pools do not have a location).
> 
>     2. Is the volume located on another cluster member? If so, and the target parameter is set, use the target parameter as the location. If so, and the target parameter is unset, use the cluster member name as the location. If not, use the server name as the location.

Before reviewing further I just wanted to check and understand this statement, as it seems over complicated to me for an access check of a volume (which in my understanding ultimately boils down to loading the volume record to get its ID before doing an access check on that resource).


Or is this just the existing logic that is used to load a specific volume already?

File: shared/entity/type.go (hunk 5 above)
Comment (markylaing) [reply shared/entity/type.go:225]: The `Target` method called here is what ignores empty or "none" locations. 

This change is so that when calling this method with e.g. `(api.StorageVolume).Location` doesn't add "target=none" to the URL if we are in standalone mode, or "target=" for remote pools.

File: lxd/storage_buckets.go (hunk 7 above)
Comment (markylaing) [reply lxd/storage_buckets.go:90]: Ah yes, in the context of fine-grained auth, the unique URL for the resource does not have a target parameter because the volume does not have a location. The request URL might have a target parameter. I'll update the comment.

Comment (markylaing) [reply lxd/storage_buckets.go:90]: Ah yes, in the context of fine-grained auth, the unique URL for the resource does not have a target parameter because the volume does not have a location. The request URL might have a target parameter. I'll update the comment.

Comment (markylaing) [reply lxd/storage_buckets.go:90]: Ah yes, in the context of fine-grained auth, the unique URL for the resource does not have a target parameter because the volume does not have a location. The request URL might have a target parameter. I'll update the comment.

**Comment by markylaing** [comment]
> > When performing an action on a single storage volume, we need to perform the following checks:
> > ```
> > 1. Is the pool containing the volume remote? If so, the `target` query parameter may have been used to target a particular cluster member, but should not be used in the URL for the volume (volumes in remote pools do not have a location).
> > 
> > 2. Is the volume located on another cluster member? If so, and the target parameter is set, use the target parameter as the location. If so, and the target parameter is unset, use the cluster member name as the location. If not, use the server name as the location.
> > ```
> 
> Before reviewing further I just wanted to check and understand this statement, as it seems over complicated to me for an access check of a volume (which in my understanding ultimately boils down to loading the volume record to get its ID before doing an access check on that resource).
> 
> Or is this just the existing logic that is used to load a specific volume already?

This is the logic that is used to figure out where to forward the request. Coming back to this I think it can probably be simplified by writing a custom query and bypassing all this complication.

The main issue is that there can be multiple volumes with the same name, just located on different members.

**Comment by markylaing** [comment]
@tomponline summarising here my remarks about this PR in the core daily stand up:

As far as fine-grained auth is concerned, we can call `CheckPermission` on a URL if that URL represents a single unique resource. We are using `cluster.GetEntityReferenceFromURL` to do this in the OpenFGA storage driver.

For storage buckets and volumes, this represents a challenge because bucket/volumes can have the same name if they are in a local pool and located on different cluster members. Additionally, a `target` query parameter can be provided if the caller wants to make the request to a specific member, regardless of the actual location of the bucket/volume.

There are a few cases to consider:
1. Volume in remote pool, target parameter provided.
2. Volume in remote pool, target parameter not provided.
3. Volume in local pool, target parameter provided.
4. Volume in local pool, target parameter not provided.

To catch all these cases, we need to*:
1. Load the storage pool and check if it is remote or not.
2. If it is remote:
    i. Strip the target parameter and use this URL in the access check.
3. If it is local:
    i. If the target parameter was provided, use it in the URL for the access check.
    ii. Otherwise, find the node that the volume is located on and use that member name in the access check.

*For all of these cases, we additionally need to check for the effective project of the bucket/volume and substitute the project parameter if it differs from that of the request.

In figuring out all of these steps, we've already done a lot of the work that would normally be done in the handler for forwarding requests to appropriate nodes. This PR adds the effective project, the storage pool, and forwarding node info (name & address) to the request context so that the handlers don't need to perform these queries again. Additionally, if the request has already been forwarded, the query for forwarding node info is skipped (we assume the other node has already correctly handled this).

I think this PR is the way forward because it overall reduces the number of queries (by skipping the remote volume check when it's a forwarded request) but does not change the queries themselves. (We can optimise these later if it becomes necessary).

**Comment by markylaing** [comment]
@tomponline have updated this to address your comments. Ready for review when you have some time. Thanks.

**Comment by tomponline** [comment]
Please can you rebase this

**Comment by markylaing** [comment]
> Please can you rebase this

Yep done.

File: lxd/storage_volumes.go [hunk 8]
```diff
@@ -2727,3 +2727,84 @@ func createStoragePoolVolumeFromBackup(s *state.State, r *http.Request, requestP
...
+		return nil, err
+	}
+
+	// If volume uses a remote storage driver and so has no explicit cluster member, then we need to check
+	// whether it is exclusively attached to remote instance, and if so then we need to forward the request to
+	// the node whereit is currently used. This avoids conflicting with another member when using it locally.
+	if err == db.ErrNoClusterMember {
+		// GetStoragePoolVolume returns a volume with an empty Location field for remote drivers.
+		var dbVolume *db.StorageVolume
+		err = s.DB.Cluster.Transaction(ctx, func(ctx context.Context, tx *db.ClusterTx) error {
```
Comment (tomponline) [inline lxd/storage_volumes.go:2861]: I believe this transaction could be avoided by moving the conditional  tx.GetStoragePoolVolume call into the earlier transaction based on the error value returned by the earlier DB calls.

File: lxd/storage_volumes.go [hunk 9]
```diff
@@ -2728,6 +2728,84 @@ func createStoragePoolVolumeFromBackup(s *state.State, r *http.Request, requestP
...
+	// particular member).
+	storagePool, err := storagePools.LoadByName(s, poolName)
+	if err != nil {
+		return err
+	}
+
+	request.SetCtxValue(r, ctxStoragePool, storagePool)
+
+	// Get the effective project.
+	actualProject, err := project.StorageVolumeProject(s.DB.Cluster, request.ProjectParam(r), volumeType)
```
Comment (tomponline) [inline lxd/storage_volumes.go:2796]: should be called effectiveProject for consistency

File: lxd/storage_volumes.go [hunk 10]
```diff
@@ -2728,6 +2728,84 @@ func createStoragePoolVolumeFromBackup(s *state.State, r *http.Request, requestP
 	return operations.OperationResponse(op)
 }
 
+const (
+	ctxStoragePool                 = "storage-pool"
+	ctxStorageVolumeRemoteNodeInfo = "storage-volume-forwarding-node-info"
+)
+
+// addStoragePoolVolumeDetailsToRequestContext extracts details from HTTP requests that pertain to a particular storage
```
Comment (tomponline) [inline lxd/storage_volumes.go:2756]: The comments for these types of function should document what they add to the context, otherwise the review/user has to read the full implementation to understand what it does, which shouldn't be necessary.

File: lxd/storage_volumes.go [hunk 11]
```diff
@@ -80,6 +80,64 @@ var storagePoolVolumeTypeCmd = APIEndpoint{
...
+// storagePoolVolumeTypeAccessHandler returns an access handler which checks the given entitlement on a storage volume.
+func storagePoolVolumeTypeAccessHandler(entitlement auth.Entitlement) func(d *Daemon, r *http.Request) response.Response {
+	return func(d *Daemon, r *http.Request) response.Response {
+		s := d.State()
+		err := addStoragePoolVolumeDetailsToRequestContext(s, r)
+		if err != nil {
+			return response.SmartError(err)
+		}
+
+		volumeName, err := url.PathUnescape(mux.Vars(r)["volumeName"])
```
Comment (tomponline) [inline lxd/storage_volumes.go:92]: Should we add these to the context in `addStoragePoolVolumeDetailsToRequestContext` as it does this same work in there too and would seem to me to make sense to separate concerns of parsing the URL into volume details rather than doing part of it twice in different places.

File: lxd/storage_buckets.go [hunk 12]
```diff
@@ -55,6 +55,55 @@ var storagePoolBucketKeyCmd = APIEndpoint{
 	Put:    APIEndpointAction{Handler: storagePoolBucketKeyPut, AccessHandler: allowPermission(entity.TypeStorageBucket, auth.EntitlementCanEdit, "poolName", "bucketName")},
 }
 
+// storageBucketAccessHandler returns an access handler that checks for the given entitlement against a storage bucket.
+// The storage pool containing the bucket and the effective project of the bucket are added to the request context for
+// later use.
+func storageBucketAccessHandler(entitlement auth.Entitlement) func(d *Daemon, r *http.Request) response.Response {
```
Comment (tomponline) [inline lxd/storage_buckets.go:61]: Why doesn't this function use an equivalent of addStoragePoolVolumeDetailsToRequestContext like storage volumes do?

File: lxd/storage_volumes.go (hunk 8 above)
Comment (markylaing) [reply lxd/storage_volumes.go]: Fixed.

File: lxd/storage_volumes.go (hunk 9 above)
Comment (markylaing) [reply lxd/storage_volumes.go]: Fixed.

File: lxd/storage_volumes.go (hunk 10 above)
Comment (markylaing) [reply lxd/storage_volumes.go]: I've added these details to the function comment now.

File: lxd/storage_buckets.go (hunk 12 above)
Comment (markylaing) [reply lxd/storage_buckets.go:61]: I've updated storage buckets to use the same pattern as for storage volumes.

File: lxd/storage_volumes.go (hunk 11 above)
Comment (markylaing) [reply lxd/storage_volumes.go]: I've now created two structs `storageVolumeDetails` and `storageBucketDetails` to hold all of this information and be stored in the request context, so we're no longer parsing any request details more than once.

File: lxd/storage_volumes.go [hunk 13]
```diff
@@ -2729,6 +2729,110 @@ func createStoragePoolVolumeFromBackup(s *state.State, r *http.Request, requestP
...
+// storageVolumeDetails contains details common to all storage volume requests. A value of this type is added to the
+// request context when addStoragePoolVolumeDetailsToRequestContext is called. We do this to avoid repeated logic when
+// parsing the request details and/or making database calls to get the storage pool or effective project. These fields
+// are required for the storage volume access check, and are subsequently available in the storage volume handlers.
+type storageVolumeDetails struct {
+	volumeName           string
+	volumeTypeName       string
+	volumeType           int
+	pool                 storagePools.Pool
+	requestProjectName   string
```
Comment (tomponline) [inline lxd/storage_volumes.go:2633]: This field seems like something we should have a common context field for, rather than embedding into every entity type specific context structure?

Also the `effectiveProjectName` could also be one that isn't specific to the entity type, but rather will be set by the middleware handler based on the type of entity in the request and the project settings.

File: lxd/storage_volumes.go [hunk 14]
```diff
@@ -2729,6 +2729,110 @@ func createStoragePoolVolumeFromBackup(s *state.State, r *http.Request, requestP
...
+	effectiveProjectName string
+	target               string
```
Comment (tomponline) [inline lxd/storage_volumes.go:2635]: These two also feel like they are not entity type specific and would be better off as their own context fields, what do you think?

File: lxd/storage_volumes.go [hunk 15]
```diff
@@ -81,6 +81,53 @@ var storagePoolVolumeTypeCmd = APIEndpoint{
...
+
+		// Regardless of whether the caller specified a target parameter, we do not add it to the authorization check if
+		// the storage pool is remote. This is because the volume in the database has a NULL `node_id`, so the URL uniquely
+		// identifies the volume without the target parameter.
+		if !details.pool.Driver().Info().Remote {
+			// If the storage pool is local, we need to add a target parameter to the authorization check URL for the
+			// auth subsystem to consider it unique.
+
+			// If the target parameter was specified, use that.
+			target = request.QueryParam(r, "target")
```
Comment (tomponline) [inline lxd/storage_volumes.go:108]: I'm a bit confused about this part.

On the one hand, you're added target to the storage volume context field in `addStoragePoolVolumeDetailsToRequestContext` , so it would seem this should be taken out of the context rather than the URL.

Also I suggested earlier that as target isn't specific to storage volumes that it shouldn't be in the storage volume context field and could be its own dedicated context field, but if we are just able to access it via the URL here, why do we need it in the context at all (in any field)?

Comment (markylaing) [reply lxd/storage_volumes.go:108]: I think on balance from your comments I will:
1. Remove `requestProject` from `storage{Volume,Bucket}Details`. We can just use `request.ProjectParam` for this. I don't think we need to dedicate a CtxKey for this since it is easy to parse without additional error handling.
2. Remove `effectiveProject` from `storage{Volume,Bucket}Details`. We will expect this to be available in the context via `request.CtxEffectiveProjectName`.
3. Remove `target` from `storage{Volume,Bucket}Details`. We can just use `request.QueryParam(r, "target")` for this. I don't think we need to dedicate a CtxKey for this since it is easy to parse without additional error handling.

Comment (tomponline) [reply lxd/storage_volumes.go:108]: sounds good!

Comment (markylaing) [reply lxd/storage_volumes.go:108]: I've updated the PR with these changes now.

File: lxd/storage_volumes.go [hunk 16]
```diff
@@ -2729,6 +2729,104 @@ func createStoragePoolVolumeFromBackup(s *state.State, r *http.Request, requestP
...
+	forwardingNodeInfo *db.NodeInfo
+}
+
+// addStoragePoolVolumeDetailsToRequestContext extracts storageVolumeDetails from the http.Request and adds it to the
+// request context with the ctxStorageVolumeDetails request.CtxKey. Additionally, the effective project of the storage
+// bucket is added to the request context under request.CtxEffectiveProjectName.
+func addStoragePoolVolumeDetailsToRequestContext(s *state.State, r *http.Request) error {
+	var details storageVolumeDetails
+	defer func() {
+		request.SetCtxValue(r, ctxStorageVolumeDetails, details)
```
Comment (tomponline) [inline lxd/storage_volumes.go:2670]: Why do you do this as a defer rather than at the end, as otherwise it maybe set to partial state?

Comment (markylaing) [reply lxd/storage_volumes.go:2670]: We expect a partial state when either the `target` parameter is set, or when the request has already been forwarded (by checking `request.CtxProtocol`). In this case the `forwardingNodeInfo` field is not set. This defer was a convenience really, to make sure it is set in the context. An error is returned if we fail to parse anything.

**Review by tomponline — APPROVED** [review]
Thanks! This pattern of extracting fields common to several API endpoints and loading the DB resource and storing them into the context seems like a good way of avoiding additional DB queries when performing access checks as well as removing code duplication in the API endpoint handlers themselves.
